arch/arm/core/swap.S

```diff
@@ -324,6 +352,47 @@ _oops:
     blx _do_kernel_oops
     pop {pc}
 
+#if CONFIG_USERSPACE
+    /*
+     * System call will setup a jump to the _do_arm_syscall function
+     * when the SVC returns via the bx lr.
+     *
+     * There is some trickery involved here because we have to preserve
+     * the original LR value so that we can return back to the caller of
+     * the SVC.
+     *
+     * On SVC exeption, the stack looks like the following:
+     * r0 - r1 - r2 - r3 - r12 - LR - PC - PSR
+     * r5 - r6 - call id - saved LR
+     *
+     */
+_do_syscall:
+    ldr r1, [r0, #24]   /* grab address of PC from stack frame */
+    str r1, [r0, #44]   /* store address to use for LR after syscall */
+    ldr r1, =_arm_do_syscall
+    str r1, [r0, #24]   /* overwrite the LR to point to _arm_do_syscall */
+
+    /* validate syscall limit, only set priv mode if valid */
```

Comment:
Second part of this comment is no longer true.

Reply:
So why bother validating here at all, and not just do it in _arm_do_syscall?

Reply:
I'll fix the comment.  So in a previous review I was doing just that, but there was a comment about making sure not to progress any further into the SVC without validating.  I was having to do the syscall check in both places because I am limited to what I can store or overwrite because of the nature of the SVC call/return.  So I opted to do it here.

Reply:
If we are going to do the priv esc regardless, I'm not sure what we are gaining by doing the check here vs in _arm_do_syscall itself (it would make more sense if we were going to oops or otherwise fail immediately w/o privilege elevation and w/o calling _arm_do_syscall at all).  The risk of only checking here is that an inconsistency could arise between this check and the actual use by _arm_do_syscall. Also not sure you want _arm_do_syscall to be a gadget for calling any function pointer in memory after the system call table, since it is fully trusting its caller to have done the validation and set SYSCALL_BAD a priori.

Reply:
Calling any function pointer in memory would only be the case if the thread making the call to the _arm_do_syscall is privileged.  If the thread was unprivileged, it'd fault at the point of dealing with the privileged stack as it would not have the permissions to access that memory region.  And this would occur before the branch to the function ptr.